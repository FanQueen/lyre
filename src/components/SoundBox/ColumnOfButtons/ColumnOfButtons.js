import React from 'react';
import styled from 'styled-components';
import { compose, withReducer, lifecycle } from 'recompose';

export const ColumnOfButtonsWrapper = styled.div`
  display: flex;
  flex-flow: column wrap;
  margin-right: 1px;
  margin-left: 1px;
`;

export const columnReducer = (state, action) => {
  switch (action.type) {
    case 'TOGGLE_SELECTED_BOX': {
      const { clickedButton } = action;
      return clickedButton === state ? false : clickedButton;
    }
    default: {
      return state;
    }
  }
};

export const ColumnOfButtons = ({ state, dispatch, colorButton }) => {
  const ColorButton = colorButton;
  const colorButtons = [1, 2, 3, 4].map(buttonNumber => (
    <ColorButton
      onClick={() =>
        dispatch({ type: 'TOGGLE_SELECTED_BOX', clickedButton: buttonNumber })
      }
      selected={buttonNumber === state}
      key={buttonNumber}
    />
  ));
  return <ColumnOfButtonsWrapper>{colorButtons}</ColumnOfButtonsWrapper>;
};

const loadTrack = (type, trackNumber, context, soundBank, URL) => {
  return new Promise((resolve, reject) => {
    const request = new XMLHttpRequest();
    request.open('GET', URL, true);
    request.responseType = 'arraybuffer';
    request.onload = () => {
      context.decodeAudioData(request.response, buffer => {
        const createNode = () => {
          const node = context.createBufferSource();
          node.buffer = buffer;
          node.connect(context.destination);
          return node;
        };
        const duration = createNode().buffer.duration;
        soundBank[type + trackNumber] = { createNode, duration };
        resolve();
      });
    };
    request.send();
  });
};

const startTrack = (type, trackNumber, context, clock, soundBank, event) => {
  if (event) {
    event.clear();
    event.bufferNode.stop();
  }
  event = clock
    .callbackAtTime(event => {
      const bufferNode = soundBank[type + trackNumber].createNode();
      bufferNode.start();
      event.bufferNode = bufferNode;
    }, context.currentTime)
    .repeat(soundBank[type + trackNumber].duration);
  return event;
};

export const ColumnOfButtonsWithState = compose(
  withReducer('state', 'dispatch', columnReducer, false)
)(ColumnOfButtons);

export default compose(
  withReducer('state', 'dispatch', columnReducer, false),
  lifecycle({
    componentWillUpdate({
      state,
      soundBank,
      audioFiles,
      context,
      clock,
      type
    }) {
      if (state) {
        if (!soundBank[type + state]) {
          const url = audioFiles[type][state - 1];
          loadTrack(type, state, context, soundBank, url).then(() => {
            this.event = startTrack(
              type,
              state,
              context,
              clock,
              soundBank,
              this.event
            );
          });
        } else {
          this.event = startTrack(
            type,
            state,
            context,
            clock,
            soundBank,
            this.event
          );
        }
      } else {
        if (this.event) {
          this.event.clear();
          this.event.bufferNode.stop();
        }
      }
    }
  })
)(ColumnOfButtons);
