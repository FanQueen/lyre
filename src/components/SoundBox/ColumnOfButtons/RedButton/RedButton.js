import styled, { css } from 'styled-components';
import { getBlinker, ButtonStyles } from '../../../../styles/ButtonStyles';

const redColor = css`#e2818a`;
const blinker = getBlinker(redColor);

const RedButton = styled.button`
  ${ButtonStyles} ${props =>
      props.selected &&
      !props.isLoading &&
      css`
        background: ${redColor};
      `} ${props =>
      props.isLoading &&
      props.selected &&
      css`
        animation: ${blinker} 1s steps(2, end) infinite;
      `};
`;

export default RedButton;
