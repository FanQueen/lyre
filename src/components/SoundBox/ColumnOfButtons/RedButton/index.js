export { default } from './RedButton.js';
