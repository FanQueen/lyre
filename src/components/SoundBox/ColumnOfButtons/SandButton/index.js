export { default } from './SandButton.js';
