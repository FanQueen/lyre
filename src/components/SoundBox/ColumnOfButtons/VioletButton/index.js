export { default } from './VioletButton.js';
