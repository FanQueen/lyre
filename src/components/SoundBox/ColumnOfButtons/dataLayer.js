export const loadTrack = (audioType, trackNumber, clock, soundBank, URL) => {
  return new Promise((resolve, reject) => {
    const request = new XMLHttpRequest();
    request.open('GET', URL, true);
    request.responseType = 'arraybuffer';
    request.onload = () => {
      const context = clock.context;
      context.decodeAudioData(request.response, buffer => {
        const duration = buffer.duration;
        const createNode = () => {
          const node = context.createBufferSource();
          node.buffer = buffer;
          node.connect(context.destination);
          return node;
        };
        soundBank[audioType + trackNumber] = { createNode, duration };
        resolve();
      });
    };
    request.send();
  });
};

export const startTrack = (audioType, trackNumber, clock, soundBank) => {
  restartTracks(clock);
  const context = clock.context;
  const event = clock
    .callbackAtTime(event => {
      const bufferNode = soundBank[audioType + trackNumber].createNode();
      bufferNode.start();
      event.bufferNode = bufferNode;
    }, context.currentTime)
    .repeat(soundBank[audioType + trackNumber].duration);
  return event;
};

export const restartTracks = clock => {
  const events = clock._events;
  const context = clock.context;
  events.forEach(event => {
    event.bufferNode.stop();
    event.schedule(context.currentTime);
  });
};

export const changeTrack = (
  selectedButton,
  soundBank,
  audioFiles,
  clock,
  audioType
) => {
  if (soundBank[audioType + selectedButton]) {
    return Promise.resolve(
      startTrack(audioType, selectedButton, clock, soundBank)
    );
  } else {
    const audioFileIndex = selectedButton - 1;
    const url = audioFiles[audioType][audioFileIndex];
    return loadTrack(audioType, selectedButton, clock, soundBank, url).then(
      () =>
        Promise.resolve(startTrack(audioType, selectedButton, clock, soundBank))
    );
  }
};

export const stopTrack = event => {
  event.clear();
  event.bufferNode.stop();
};

export const newTrack = (
  track,
  clickedButton,
  selectedButton,
  soundBank,
  audioFiles,
  clock,
  audioType
) => {
  if (track) {
    track.then(stopTrack);
  }
  if (clickedButton === selectedButton) {
    return null;
  }
  return changeTrack(clickedButton, soundBank, audioFiles, clock, audioType);
};
