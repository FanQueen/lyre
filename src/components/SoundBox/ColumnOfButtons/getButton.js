import styled, { css } from 'styled-components';
import { ButtonStyles, getBlinker } from '../../../styles/ButtonStyles';

const getButton = color => {
  const blinker = getBlinker(color);
  return styled.button`
    ${ButtonStyles} ${props => {
        if (props.selected) {
          return props.isLoading
            ? css`
                animation: ${blinker} 1s steps(2, end) infinite;
              `
            : css`
                background: ${color};
              `;
        }
      }};
  `;
};

export default getButton;
