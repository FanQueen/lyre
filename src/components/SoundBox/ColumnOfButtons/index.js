export { default } from './ColumnOfButtons.js';
