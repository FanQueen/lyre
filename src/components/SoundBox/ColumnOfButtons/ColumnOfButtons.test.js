import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import ColumnOfButtons from './';
import RedButton from './RedButton';
import WAAClock from 'waaclock';
import audioFiles from '../audioFiles';

describe('ColumnOfButtons with state', () => {
  it('renders ColumnOfButtons correctly', () => {
    const renderedColumnOfButtons = mount(
      <ColumnOfButtons colorButton={RedButton} />
    );
    expect(toJson(renderedColumnOfButtons)).toMatchSnapshot();
  });

  it('renders ColumnOfButtons with 4 color buttons', () => {
    const renderedColumnOfButtons = mount(
      <ColumnOfButtons colorButton={RedButton} />
    );
    expect(renderedColumnOfButtons.find(RedButton)).toHaveLength(4);
  });

  it('ColumnOfButtons changes state on select/unselect button', () => {
    const context = new window.AudioContext();
    const clock = new WAAClock(context);
    const soundBank = {};
    clock.start();
    const renderedColumnOfButtons = mount(
      <ColumnOfButtons
        soundBank={soundBank}
        audioFiles={audioFiles}
        audioType="arp"
        clock={clock}
        colorButton={RedButton}
      />
    );
    const firstSimpleBox = renderedColumnOfButtons.find(RedButton).first();
    expect(
      renderedColumnOfButtons.state().stateValue.selectedButton
    ).toBeFalsy();
    firstSimpleBox.simulate('click');
    expect(renderedColumnOfButtons.state().stateValue.selectedButton).toEqual(
      1
    );
    firstSimpleBox.simulate('click');
    expect(
      renderedColumnOfButtons.state().stateValue.selectedButton
    ).toBeFalsy();
  });
});
