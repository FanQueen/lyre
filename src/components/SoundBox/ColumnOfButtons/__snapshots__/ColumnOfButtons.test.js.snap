// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnOfButtons with state renders ColumnOfButtons correctly 1`] = `
<withReducer(Component)
  colorButton={[Function]}
>
  <Component
    colorButton={[Function]}
    dispatch={[Function]}
    state={
      Object {
        "selectedButton": null,
      }
    }
  >
    <styled.div>
      <div
        className="sc-bdVaJa cMfvsI"
      >
        <styled.button
          key="1"
          onClick={[Function]}
          selected={false}
        >
          <button
            className="sc-bwzfXH kGMlao"
            onClick={[Function]}
            selected={false}
          />
        </styled.button>
        <styled.button
          key="2"
          onClick={[Function]}
          selected={false}
        >
          <button
            className="sc-bwzfXH kGMlao"
            onClick={[Function]}
            selected={false}
          />
        </styled.button>
        <styled.button
          key="3"
          onClick={[Function]}
          selected={false}
        >
          <button
            className="sc-bwzfXH kGMlao"
            onClick={[Function]}
            selected={false}
          />
        </styled.button>
        <styled.button
          key="4"
          onClick={[Function]}
          selected={false}
        >
          <button
            className="sc-bwzfXH kGMlao"
            onClick={[Function]}
            selected={false}
          />
        </styled.button>
      </div>
    </styled.div>
  </Component>
</withReducer(Component)>
`;
